lib/Basic/Cache.cpp

```diff
@@ -78,14 +78,16 @@ void CacheImpl::setAndRetain(void *Key, void *Value, size_t Cost) {
   DefaultCacheKey CKey(Key, &DCache.CBs);
   auto Entry = DCache.Entries.find(CKey);
   if (Entry != DCache.Entries.end()) {
-    if (Entry->second == Value)
-      return;
     DCache.CBs.keyDestroyCB(Entry->first.Key, nullptr);
-    DCache.CBs.valueReleaseCB(Entry->second, nullptr);
+    if (Entry->second != Value) {
+      DCache.CBs.valueReleaseCB(Entry->second, nullptr);
+      DCache.CBs.valueRetainCB(Value, nullptr);
+    }
     DCache.Entries.erase(Entry);
+  } else {
+    DCache.CBs.valueRetainCB(Value, nullptr);
```

Comment:
This logic is getting kind of hard to follow, and I'm not sure how safe it is to `keyDestroy` but leave the key in the map temporarily and then call `operator[]` which will compare against that dead key.  How about something like this (untested):

```c++
if (Entry == DCache.Entries.end()) {
  DCache.CBs.valueRetainCB(Value, nullptr);		
  DCache.Entries[CKey] = Value;
  return;
}

// Destroy the original key, which must be == the new key.
std::swap(Entry->first.Key, Key);
DCache.CBs.keyDestroyCB(Key);

// Replace the value, if necessary.
if (Entry->second != Value) {
  DCache.CBs.valueRetainCB(Value, nullptr);	
  std::swap(Entry->second, Value);
  DCache.CBs.valueReleaseCB(Value, nullptr);
}
```

Reply:
Thanks, that's a lot clearer and it works.